src/errors/errors.ts

```diff
@@ -1,153 +1,209 @@
-    this.name = 'NotFoundError';
-  }
+export class FailedPreconditionError extends SdkError {
+  _errorCode = MomentoErrorCode.FAILED_PRECONDITION_ERROR;
+  _messageWrapper =
+    "System is not in a state required for the operation's execution";
 }
 
 /**
  * Cache Service encountered an unexpected exception while trying to fulfill the request
  */
-export class InternalServerError extends CacheServiceError {
-  constructor(message: string, stack: string) {
-    super(message);
-    this.name = 'InternalServerError';
-    this.stack = stack;
-  }
+export class InternalServerError extends SdkError {
+  _errorCode = MomentoErrorCode.INTERNAL_SERVER_ERROR;
+  _messageWrapper =
+    'An unexpected error occurred while trying to fulfill the request; please contact us at support@momentohq.com';
 }
 
 /**
- * Insufficient permissions to perform an operation on Cache Service
+ * Represents errors thrown when invalid parameters are passed to the Momento Cache
  */
-export class PermissionError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'PermissionError';
-  }
+export class InvalidArgumentError extends SdkError {
+  _errorCode = MomentoErrorCode.INVALID_ARGUMENT_ERROR;
+  _messageWrapper = 'Invalid argument passed to Momento client';
 }
 
 /**
- * Error when authentication with Cache Service fails
+ * Error when calls are throttled due to request limit rate
  */
-export class AuthenticationError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'AuthenticationError';
-  }
+export class LimitExceededError extends SdkError {
+  _errorCode = MomentoErrorCode.LIMIT_EXCEEDED_ERROR;
+  _messageWrapper =
+    'Request rate exceeded the limits for this account.  To resolve this error, reduce your request rate, or contact us at support@momentohq.com to request a limit increase';
 }
 
 /**
- * Error when an operation with Cache Service was cancelled
+ * Error that occurs when trying to get a cache that doesn't exist. To resolve, make sure that the cache you are trying
+ * to get exists. If it doesn't create it first and then try again
  */
-export class CancelledError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'CancelledError';
-  }
+export class NotFoundError extends SdkError {
+  _errorCode = MomentoErrorCode.NOT_FOUND_ERROR;
+  _messageWrapper =
+    'A cache with the specified name does not exist.  To resolve this error, make sure you have created the cache before attempting to use it';
 }
 
 /**
- * Error when calls are throttled due to request limit rate
+ * Insufficient permissions to perform an operation on Cache Service
  */
-export class LimitExceededError extends CacheServiceError {
-  constructor(message: string) {
-    super(message);
-    this.name = 'LimitExceededError';
-  }
+export class PermissionError extends SdkError {
+  _errorCode = MomentoErrorCode.PERMISSION_ERROR;
```

Comment:
Just curious, do we currently ever throw this error? I have never seen it before

Reply:
I imagine so. There is a corresponding gRPC status that gets mapped to this error type.